### PR TITLE
Move NPM package files out of /lib ; affects [resharper nuget myget dub chocolatey github]

### PR DESCRIPTION
**Pull request by chris48s**
The aim of this PR is to move all the files relating to the gh-badges package into a new location and give the package its own `package.json` and documentation while making as few other changes as possible.

The accomplishes several things:

* We can distribute a package which only requires the necessary dependencies for the library (at the moment we force our users to install a bunch of other dependencies needed by shields.io but not gh-badges)
* It is easier to provide more relevant documentation for the NPM package
* It is step 1 on the road to further reducing the coupling between this package and the rest of the codebase

I've made the decision to keep all the npm `devDependencies` and `scripts` (for testing etc) in the repo root. There are several reasons for this:

* `npm install "file:gh-badges/"` doesn't install the `devDependencies` of the child package and its (prod) dependencies go into the parent `node_modules`.
* We'll need to amend our dependabot config to track/update dependencies across 2 `package.json`s. We're going to get 2 PRs to review/merge for any packages which exist in both `package.json`s. I'm keen to minimise that overhead. As it stands, `is-css-color` is the only dependency we need to have in both `package.json`s.

I think this is the right choice for the way we organise things now, but if we want to move to a position where we store the package in a different repo we would need to change this.

A couple of other notes:

* I haven't moved the logos dir into gh-badges. The logos dir _was_ included in the package in 1.1.1, 1.3.0 and 2.0.0 but having looked into this a bit more, I don't think we've ever advertised/documented that they exist or documented `logo` as a valid format key. I think moving forwards, we should not include the logos in the package.
* There is some further work that I would like to do around what we export in the package and how we import/require elsewhere in the application, but I don't want to clutter this PR more than we need to - that can happen in a future PR.
* This PR modifies the service tests for powershellgallery but I haven't put [powershellgallery] in the title because they're failing for an unrelated reason. I think the temporary endpoint we switched to (see https://github.com/badges/shields/issues/2078 ) has gone away. Lets not get bogged down in that issue in this PR.

### CODE REVIEW

File: gh-badges/index.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 'use strict'
```
[the remark this answers is not in the document]
Comment (chris48s) [reply gh-badges/index.js]: Are you thinking we should move the `BadgeFactory` class into `/lib` and strip down `index.js` to just:


```js
const BadgeFactory = require('./lib/whatever')

module.exports = {
  BadgeFactory,
}
```

or do you want to move it all into `/lib` and set `"main": "lib/whatever.js"` in `package.json`?

I'd prefer the first personally. I intentionally moved this to `index.js` in the root as it is standard to expose your public interface there. I'm in favour of going with convention on this unless there's a strong reason not to.

**Comment by chris48s** [comment]
> The only medium thing is the name of the folder. While gh-badges is the historical name of the npm package, it’s not very descriptive and I think does not convey to a new developer what to expect in the folder. I’d suggested some alternate names for the npm package. What do you think about using one of those names for this folder? If you want it to match the package name, we could deprecate the old package and re-release under a new name. In that case we should pick a name that’s available in npm.

Ideally the directory name should mirror the package name so if we're going to continue publishing at https://www.npmjs.com/package/gh-badges then the dir should be called gh-badges. If we're going to move it to a new name in NPM then we should call the directory whatever the new package name is. I think its confusing if we change the dir name but continue releasing with gh-badges as the package name.

I do agree that gh-badges is not the most descriptive name. It is not what I would choose today, but there is history. That said, I think I am in favour of sticking with it for now. It seems a bit odd not to put out a release for 2 years, then pop up and release 2.0 and then immediately deprecate the package a week later.

Perhaps another approach would be to call the dir 'package' or something - then we can change the package name at a later date without moving it?


> it would be helpful to write a couple sentences in CONTRIBUTING.md

Yeah - good point. I'll have a look at the contributing docs too


> if your intent is that contents of gh-badges is imported using relative imports rather than require('gh-badges') it'd be good to mention that.

I'm not entirely sure what the best approach is here. I think if we want to treat this more like a standalone package, we want to move away from relative imports but I think that will also mean we need to expose more of the guts of the module as public interface (e.g: if we want to use `make-badge-test-helpers` in other tests in `/lib` we need to expose it to package users too, but that's maybe not necessarily a massively useful thing to do). Maybe the answer is a mix. Thinking about this issue some more is what I was getting at when I said:

> There is some further work that I would like to do around what we export in the package and how we import/require elsewhere in the application

I've just done everything with relative imports for the moment as I wanted to avoid changing too many things all at once in the same PR, but I think the next thing (or one of the next things) we should do is look at that issue.

Would be interested in your thoughts on the tradeoffs involved in this..


> I'm not sure whether I agree or not. I do think the simple-icons support probably should be in gh-badges, perhaps as an optional peer dependency. That could happen in a follow-on.

Lets not address SimpleIcons in this PR. Do you think we should move the _custom_ logos dir in this PR and keep them a part of the package?

File: gh-badges/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply gh-badges/index.js]: I was thinking `lib/index.js`, though I certainly don't feel strongly about it.

File: Makefile [hunk 2]
```diff
@@ -12,7 +12,7 @@ all: website favicon test
 
 favicon:
 	# This isn't working right now. See https://github.com/badges/shields/issues/1788
-	node lib/badge-cli.js '' '' '#bada55' .png > favicon.png
+	node gh-badges/lib/badge-cli.js '' '' '#bada55' .png > favicon.png
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply Makefile]: `favicon` should be removed from the `all` target too.

File: package.json [hunk 3]
```diff
@@ -74,7 +70,7 @@
     "prettier-check": "prettier-check \"**/*.js\"",
     "danger": "danger",
     "test:js:frontend": "NODE_ENV=mocha mocha --require @babel/polyfill --require @babel/register \"frontend/**/*.spec.js\"",
-    "test:js:server": "HANDLE_INTERNAL_ERRORS=false mocha \"*.spec.js\" \"lib/**/*.spec.js\" \"services/**/*.spec.js\"",
+    "test:js:server": "HANDLE_INTERNAL_ERRORS=false mocha \"*.spec.js\" \"lib/**/*.spec.js\" \"gh-badges/**/*.spec.js\" \"services/**/*.spec.js\"",
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply package.json]: Awesome.

In `.circleci/config.yml`, the new target needs to be added to `main` and `main@node-latest`

**Comment by paulmelnikow** [comment]
I proposed a solution that simplifies font measuring in #2311. Let's get this in and then we can pick that up.

**Comment by chris48s** [comment]
Hi sorry I didn't get a chance to check in on this yesterday. Thanks for responding to #2312 and adding some extra docs here.


I think I'm in favour of:

> Plan to rename the package in a couple months, and move the files again when we do that. (i.e. folder temporarily uses legacy name)

One other thing to consider is that we're now we've started talking about stuff like changing the text measurer, supporting SVG only, changing the exported classes/functions, we might find we want to make another non backwards-compatible change in the medium-term future. That would require a major version bump to 3.0. That might also represent a natural milestone where changing the package name and deprecating would make sense. I don't think renaming the directory will be a complicated job when we do this.

> Honestly when it comes to almost every decision about importing, I think the easiest solution is to refactor and rewrite these kludges away

:+1:

**Pull request merged**